drivers/led_strip/ws2812_rpi_pico_pio.c

```diff
@@ -0,0 +1,250 @@
+/*
+ * Copyright (c) 2023 TOKITA Hiroshi
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/led_strip.h>
+#include <zephyr/drivers/misc/pio_rpi_pico/pio_rpi_pico.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/dt-bindings/led/led.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/kernel.h>
+
+#define LOG_LEVEL CONFIG_LED_STRIP_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(ws2812_rpi_pico_pio);
+
+struct rpi_pico_pio_ws2812_config {
+	const struct device *piodev;
+	const struct pinctrl_dev_config *const pcfg;
+	const struct pio_program program;
+};
+
+#ifdef CONFIG_DT_HAS_WORLDSEMI_WS2812_RPI_PICO_PIO_ENABLED
```

Comment:
> Well, is there a point of having this whole driver enabled at all if this DT property is missing ? Without, the driver only adds a PIO program but nothing more...

Yes, it does nothing. But I think it is good to compile the PIO part if enabled in dts from the consistency viewpoint.
In such cases, This part causes a compiler warning.

> This comment leads back to my previous question: why are we now handling two compatibles in this driver ?


It is what I wrote in the previous comment. Because PIO and LED strip relation are not 1:1. It is better not to combine into one. But it is used only at the same time. So I think it may be better not to split file.

Reply:
2023/11/19

Splitting PIO part into rpi_pico_pio_ws2812.c.